integration/network_partition_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/testutil"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+func TestNetworkPartition5NodeLeaderInMinority(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	clus := NewClusterV3(t, &ClusterConfig{Size: 5})
+	defer clus.Terminate(t)
+
+	leadIdx1 := clus.WaitLeader(t)
+
+	// minority: leader, follower / majority: follower, follower, follower
+	minority := []int{leadIdx1, (leadIdx1 + 1) % 5}
+	majority := []int{(leadIdx1 + 2) % 5, (leadIdx1 + 3) % 5, (leadIdx1 + 4) % 5}
+	tgts := make([]*member, len(majority))
+	for i, idx := range majority {
+		tgts[i] = clus.Members[idx]
+	}
+
+	// cut connections from minority to majority
+	for _, idx := range minority {
+		clus.Members[idx].PausePartition(t, tgts)
+	}
+
+	// minority with leader can't reach quorum, so revert back follower
+	clus.Members[leadIdx1].WaitLeaderChange(t)
+
+	// new leader must be from majority
+	donec := make(chan struct{})
+	tm := clus.Members[majority[0]].s.Cfg.ReqTimeout()
+	go func() {
+		defer close(donec)
+		clus.waitLeader(t, tgts)
+	}()
+	select {
+	case <-donec:
+	case <-time.After(tm):
+		t.Fatalf("leader election took too long (%v)", tm)
+	}
+
+	// write to majority to make it more up-to-date
+	kvc := toGRPC(clus.Client(majority[0])).KV
+	for i := 0; i < 5; i++ {
+		reqput := &pb.PutRequest{Key: []byte(fmt.Sprintf("foo%d", i)), Value: []byte("bar")}
+		if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+			t.Fatalf("#%d: couldn't put key (%v)", i, err)
+		}
+	}
+
+	// recover network partition
+	for _, idx := range minority {
+		clus.Members[idx].ResumePartition(t, tgts)
+	}
+
+	// ensure leader in majority still holds
+	leadIdx2 := clus.WaitLeader(t)
+	inMajority := false
```

Comment:
this is not necessary. waitleader should already ensure everyone agrees on the same leader, no?